### PR TITLE
update operator-framework scenario for microservices

### LINKED ISSUE #82: Implement Microservice example for Helm Operator

### PR DESCRIPTION
**Pull request by joshgav**
This fixes #82 and adds an e2e test that builds, deploys, and exercises the chart-based operator, then verifies the custom PodtatoHeadApp resource works as expected.

I used [Operator Lifecycle Manager](https://olm.operatorframework.io/) in this revision rather than directly deploying the controller/operator and CRDs, but I put basic instructions for the latter in the README and suggest we fully test and document that practice too.

### CODE REVIEW

**Comment by joshgav** [comment]
Note - tests run on the PR failed at first because PR tests don't have access to push new container images. I pushed the same commit directly to this repo as a branch to trigger writing of the needed images and then re-ran the tests for the PR and they now pass.

Per this, is there a way to enable repo admins to approve and allow writing repo packages in PR tests? We'd want to ensure all the images created in the PR use new tags (e.g., `0.2.2`) and wouldn't conflict with released images, which I did here. It would be helpful to be able to grant write access for PR tests after an initial review because other people won't be able to push another branch like me (I'm a repo admin). I'll open an issue.

File: .github/workflows/microservices-build-tests.yaml [hunk 1]
```diff
@@ -50,5 +50,7 @@ jobs:
     steps:
       - name: Checkout
         uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
       - name: Test services
         run: make test-microservices
```
Comment (awgreene) [inline .github/workflows/microservices-build-tests.yaml:56]: Nit: Newline. Applies everywhere.

File: delivery/operator/README.md [hunk 2]
```diff
@@ -0,0 +1,260 @@
+# Deliver with an operator
+
+Here's how to deliver podtato-head using the [Operator
+Framework](https://operatorframework.io) and a Helm chart-based operator built
+with its [Operator SDK](https://sdk.operatorframework.io). [Operator Lifecycle
```
Comment (awgreene) [inline delivery/operator/README.md:5]:
```suggestion
Framework](https://operatorframework.io) as a Helm chart-based operator built
with the [Operator SDK](https://sdk.operatorframework.io). [Operator Lifecycle
```

File: delivery/operator/README.md [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+      - ghcr
+    displayName: podtato-head-catalog
+    publisher: podtato-head
+    updateStrategy:
+        registryPoll:
+            interval: 10m
+EOF
+```
+
+Next declare a subscription to the podtato-head-operator in the `operators` namespace:
```
Comment (awgreene) [inline delivery/operator/README.md:119]: I prefer creating a subscription as you've done here, but I believe that the SDK [provides a command](https://sdk.operatorframework.io/docs/olm-integration/quickstart-bundle/) to create install an operator which may be worth considering.

File: delivery/operator/README.md [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+    name: podtato-head-operator
+    source: podtato-head-catalog
+    sourceNamespace: olm
+    startingCSV: podtato-head-operator.v${operator_version}
+EOF
+```
+
+The subscription triggers creation of a ClusterServiceVersion (CSV) resource its
+namespace, followed by an InstallPlan that deploys Deployments, Service
+Accounts, and other resources associated with the operator.
```
Comment (awgreene) [inline delivery/operator/README.md:142]: This is slightly off. Creating a Subscription causes the Catalog Operator to generate an installplan which records a list of resources that need to be installed in the `status.plan` array, the CSV is included in this list. The OLM operator will then reconcile the InstallPlan, creating each resource listed.

File: delivery/operator/README.md [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+
+Watch for the successful provisioning of the operator deployment:
+
+```bash
+kubectl get deployments --namespace operators --watch
```
Comment (awgreene) [inline delivery/operator/README.md:147]: Alternatively, users could watch for the podtato-head CSV to enter the `succeeded` phase.

**Review by awgreene — COMMENTED** [review]
Great work on this @joshgav! I'm a developer on the OLM team and the steps centered on deploying your operator with OLM seem correct. In full transparency, I did not have time to spin up a cluster and perform the commands myself.

It may be worthwhile to have someone from the operator-sdk team review as well, but this looks good to me.

File: delivery/operator/README.md (hunk 5 above)
Comment (joshgav) [reply delivery/operator/README.md:147]: Added, thank you.

File: delivery/kustomize/base/deployment-hat.yaml [hunk 6]
```diff
@@ -6,7 +6,7 @@ metadata:
   labels:
     app.kubernetes.io/name: podtato-head
     app.kubernetes.io/component: hat
-    app.kubernetes.io/version: '0.2.1'
+    app.kubernetes.io/version: '0.2.3'
     app.kubernetes.io/managed-by: kustomize
 spec:
   replicas: 1
```
Comment (camilamacedo86) [inline delivery/kustomize/base/deployment-hat.yaml:12]: This shows that you did not use operator-sdk to built the operator. Am I right?
WDYT about use the tool?

File: delivery/operator/README.md [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+
+export IMAGE_TAG_BASE=ghcr.io/${github_user}/podtato-head/operator
+export IMG=${IMAGE_TAG_BASE}:latest
+
+make docker-build docker-push
+make bundle bundle-build bundle-push
+make catalog-build catalog-push
+```
+
+### Deploy the operator
```
Comment (camilamacedo86) [inline delivery/operator/README.md:88]: By using operator-sdk to do that you could just tun the target `make deploy IMG=` or use operator-sdk run bundle <bundle-path> to check it with the OLM integration

File: delivery/kustomize/base/deployment-hat.yaml (hunk 6 above)
Comment (joshgav) [reply delivery/kustomize/base/deployment-hat.yaml:12]: Hi @camilamacedo86, thanks for taking a look. There are a few different tools illustrated in this repo in the `./delivery` dir, all delivering the same system from `podtato-head-microservices`. In this PR I used operator-sdk in the test scripts and README to pick up the Helm chart from the example Helm scenario in `./delivery/chart` and generate and deploy a Helm operator based on it.

I didn't actually check anything in that was generated by the SDK, it's used statelessly you might say. What do you think of that?

File: delivery/operator/README.md (hunk 7 above)
Comment (joshgav) [reply delivery/operator/README.md:88]: Thanks! I'll expand the README to mention that as a shortcut.

File: delivery/operator/README.md (hunk 3 above)
Comment (joshgav) [reply delivery/operator/README.md:120]: @awgreene on review I think you may have been referring to the `make deploy` target [mentioned below](https://github.com/podtato-head/podtato-head/pull/144#discussion_r800913922). There's a section later in the doc which talks about that option, I updated that section's title to make it clearer.

Thanks!

File: delivery/operator/README.md (hunk 7 above)
Comment (joshgav) [reply delivery/operator/README.md]: @camilamacedo86 there's a section later in the doc which briefly walks through the make commands including `make deploy`. I updated that section's title to make it clearer. Thanks!

**Pull request merged**